gitclone/checkout.go

```diff
@@ -53,49 +59,100 @@ type checkoutStrategy interface {
 // X: required parameter
 // !: used to identify checkout strategy
 // _: optional parameter
-// |==================================================================================|
-// | params\strat| commit | tag | branch | manualMR | manualPR | autoMerge | autoDiff |
-// | commit      |  X  !  |     |        |  X       |          |           |          |
-// | tag         |        |  X !|        |          |          |           |          |
-// | branch      |  _     |  _  |  X !   |  X       |  X       |  X        |          |
-// | branchDest  |        |     |        |  X       |  X       |           |  X       |
-// | PRRepoURL   |        |     |        |      !   |  X !     |    !      |    !     |
-// | PRID        |        |     |        |          |          |           |    !     |
-// | mergeBranch |        |     |        |          |          |  X !      |          |
-// |==================================================================================|
-
-func selectCheckoutMethod(cfg Config) CheckoutMethod {
-	isPR := cfg.PRRepositoryURL != "" || cfg.PRMergeBranch != "" || cfg.PRID != 0
+// |==========================================================================================================================|
+// | params\strat| commit | tag | branch | manualMR | manualPR | autoMerge | autoDiff | noMergeHeadBranch | noMergeForkBranch |
+// | commit      |  X  !  |     |        |  X       |          |           |          |                   |                   |
+// | tag         |        |  X !|        |          |          |           |          |                   |                   |
+// | branch      |  _     |  _  |  X !   |  X       |  X       |  X        |          |                   |  X                |
+// | branchDest  |        |     |        |  X       |  X       |           |  X       |                   |                   |
+// | PRRepoURL   |        |     |        |      !   |  X !     |    !      |    !     |                   |  X !              |
+// | PRID        |        |     |        |          |          |           |    !     |                   |    !              |
+// | mergeBranch |        |     |        |          |          |  X !      |          |    !              |                   |
+// | headBranch  |        |     |        |          |          |           |          |  X !              |                   |
+// |==========================================================================================================================|
+
+func selectCheckoutMethod(cfg Config, patch patchSource) (CheckoutMethod, string) {
```

Comment:
This function is doing more than the name suggests. Instead of returning the diff file in a second parameter, we could let the caller retrieve the patch file if the function returns `CheckoutPRDiffFileMethod`. As far as I could tell in all other cases we simply return an empty string. How does that sound?

Reply:
The intention was to detect early if a diff file is unavailable and possibly fall back to an other method.